Support defaults in themes

viebel opened the pull request:
Allow the theme data to include default values using `:theme-defaults` key. This key is `dissoc`ed when the theme data is passed to codox.

I need this feature for the [klipse codox theme](https://github.com/viebel/codox-klipse-theme) in order to allow the user to define the klipse selector - but I want to keep the possibility to have a default value.

This is how the theme data will look like for klipse codox theme with the default values:

```clojure
{:theme-defaults {:klipse/selector ".klipse"}
 :transforms
 [[:body]
  [:prepend
   [:link
    {:rel "stylesheet"
     :type "text/css"
     :href "https://storage.googleapis.com/app.klipse.tech/css/codemirror.css"}]
   [:div {:style "visibility: hidden;"}
    [:div {:class "klipse"
           :data-external-libs :klipse/external-libs}
     :klipse/require-statement]]]
  [:body]
  [:append
   [:script
    ("window.klipse_settings = {
    selector: " "'" :klipse/selector "',
    codemirror_options_in: {
    lineWrapping: true,
    autoCloseBrackets: true
    },
    codemirror_options_out: {
    lineWrapping: true
    }
    };" )]
 [:script {:src "https://storage.googleapis.com/app.klipse.tech/plugin/js/klipse_plugin.js"}]]]}
```

weavejester commented:
Can you change the key from `:theme-defaults` to just `:defaults`, and change the commit message to:

    Add default values to theme variables

viebel commented:
done

On Fri, 11 Nov 2016 at 00:06 James Reeves <notifications@github.com> wrote:

> Can you change the key from :theme-defaults to just :defaults, and change
> the commit message to:
>
> Add default values to theme variables
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/codox/pull/149#issuecomment-259821567>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6VPiGG_8zI_Azixp1ppwa265AAB2Bzks5q85VUgaJpZM4KvKSA>
> .
>

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -454,9 +454,11 @@
       (assert (map? params) "Theme parameters must be a map")
-      (walk/postwalk #(if (keyword? %) (params % %) %) theme-data))
+      (-> (walk/postwalk #(if (keyword? %) (params % (defaults % %)) %) theme-data)
+          (dissoc :theme-defaults)))
```
You dissoc `:theme-defaults` here, rather than `:defaults`. Is there a reason for dissociating it in the first place?

viebel commented:
Oh. Sorry for the typo.
Fixed now.
I am dissoc'ing in order not to pollute the global codox object.

viebel commented on codox/src/codox/writer/html.clj:
```diff
@@ -454,9 +454,11 @@
       (assert (map? params) "Theme parameters must be a map")
-      (walk/postwalk #(if (keyword? %) (params % %) %) theme-data))
+      (-> (walk/postwalk #(if (keyword? %) (params % (defaults % %)) %) theme-data)
+          (dissoc :defaults)))
```
nice catch! updated

weavejester commented:
Looks great. Squash your commits and I'll merge it in.

viebel commented:
Done.
Did you know that you could squash by yourself directly in github?

On Thu, 17 Nov 2016 at 17:10 James Reeves <notifications@github.com> wrote:

> Looks great. Squash your commits and I'll merge it in.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/codox/pull/149#issuecomment-261306575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6VPg3gQoiAVPeFQgj9SxbFVJc0olOtks5q_IoNgaJpZM4KvKSA>
> .
>

weavejester commented:
Thanks.

I thought squashed commits in Github would be appended to master without a merge?

weavejester commented:
Just been testing the code. If the theme is specified as a keyword and not a vector, the defaults don't work.

It's also a good idea to add defaults to the example theme so that we can test it. Perhaps even a profile in the example project file as well.

viebel commented:
1. Squashed commits in Github are appended to master with a merge.
2. I fixed the bug you mentioned and also another one: when theme has not defaults, there was a NullPointerException.

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -454,10 +454,20 @@
+    (if-let [defaults (:defaults theme-data)]
+      (do
+        (assert (map? defaults) "Theme defaults when present must be a map")
+        (->> (dissoc theme-data :defaults)
```
There's a lot of repetition here. Maybe something more like:

```clojure
(defn- insert-params [theme-data theme]
  (let [params   (if (vector? theme) (second theme) {})
        defaults (:defaults theme-data {})]
    (assert (map? params) "Theme parameters must be a map")
    (assert (map? defaults) "Theme defaults must be a map")
    (->> (dissoc theme-data :defaults)
         (walk/postwalk #(if (keyword? %) (params % (defaults % %)))))))
```

viebel commented:
Good idea  @weavejester.

Done.

I have not squashed the commits.

Let me know if you can't do it by yourself on github.

weavejester commented:
Thanks! I checked out GitHub's squash option in another project, and I was right that they're appended to master without a merge commit. This means we still need to squash manually.

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -453,11 +453,12 @@
-      (assert (map? params) "Theme parameters must be a map")
-      (walk/postwalk #(if (keyword? %) (params % %) %) theme-data))
-    theme-data))
+  (let [params   (if (vector? theme) (second theme) {})
```
I've tested this a little. I think this condition should be:

```clojure
(if (vector? theme) (or (second theme) {}) {})
```

Otherwise it doesn't account for themes like `[:test]`, that are vectors but lack a second argument.

weavejester replied:
I think that's the only change left we need.

viebel commented:
Nice catch!

Fixed and squashed.

Will you be able to deploy a new version soon? 
I need it them defaults - for klipse integration in `test.check`.

weavejester commented:
Released as 0.10.2.

viebel commented:
Cool.
Thanks for making it so quickly!

Codox really rocks!

On Mon, 21 Nov 2016 at 22:59 James Reeves <notifications@github.com> wrote:

> Released as 0.10.2.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/weavejester/codox/pull/149#issuecomment-262065211>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6VPg4r-A2Pkzx7JiF56az1C19-4w-qks5rAgYZgaJpZM4KvKSA>
> .
>
